_source/_posts/2023-07-28-oidc_workshop.md

```diff
@@ -0,0 +1,505 @@
+          return;
+        }
+      }
+```
+Don't change the sample app's logic for when the Sign In button submits both a username and a password, because users without OIDC will still provide both to log in.
+
+Since the `onUsernameEnteredFn` will be implemented in the `authState` component, make sure to import it to the Signin function in `signin.tsx`, along with the `onAuthenticateFn`: 
+
+```
+  const { onAuthenticateFn, onUsernameEnteredFn } = useAuthState();
+```
+
+The `onUsernameEnteredFn` code doesn't exist yet, but you'll write it in the next step!
+
+## Start OpenID Flow
+
+To complete the frontend changes, implement the `onUsernameEnteredFn` in `authState.tsx`. This will be similar to `onAuthenticateFn`, which is already provided in the sample code. 
+
+Declare `onUsernameEnteredFn` in the `AuthContextType` Type: 
+
+```
+  onUsernameEnteredFn: (username: string) => Promise<number|null>;
+```
+
+Add it to the `defaultAuthContext`: 
+
+```
+  onUsernameEnteredFn: async () => null,
+```
+Implement `onUsernameEnteredFn` by modifying a copy of `onAuthenticateFn` in `authState.tsx`. Instead of using the `/api/signin` endpoint on the app's backend, you'll use the `/api/openid/check` endpoint:  
+
+```
+  const onUsernameEnteredFn = async (username: string) => {
+    const url = `/api/openid/check`;
+    try {
+        const res = await fetch(url, {
+          method: 'POST',
+          headers: {
+            'Content-Type': 'application/json'
+          },
+          body: JSON.stringify({ username })
+        });
+
+        const {org_id} = await res.json();
+
+        return org_id;
+      } catch (error: unknown) {
+        console.error(error);
+      }
+
+    return null;
+  }
+```
+
+And finally, make sure that `onUsernameEnteredFn` is included in the provider context returned from `onRevokeAuthFn`: 
+
+```
+  return <AuthContext.Provider value={{ authState, onAuthenticateFn, onUsernameEnteredFn, onRevokeAuthFn, userIsAuthenticatedFn }}>{children}</AuthContext.Provider>;
+```
+
+From the perspective of the app's frontend, these are all the changes that matter. Supporting OIDC will not change how the frontend communicates with the backend through sessions. 
+
+The frontend is now ready to use the backend's `/api/openid/check` endpoint to investigate whether a user belongs to an OIDC org, and then hit the org's custom `/api/openid/start/${org_id}` endpoint to initiate the login flow if an org is found! Your next step is to add these endpoints to the application's backend. 
+
+# Backend Changes
+
+The application's backend will use a library to handle most of the OIDC authentication process. After authenticating, the user will be redirected back to the app just as they are when logging in with a password.
+
+## Install Passport OIDC Library
+
+Since the sample app is using Passport for authentication, the Passport OIDC library will offer seamless integration. 
```

Comment:
```suggestion
Since the sample app is already using Passport for password authentication, the Passport OIDC library will offer seamless integration with the OpenID Connect provider.
```